installer/scripts/agent_topology_request.rb

```diff
@@ -0,0 +1,124 @@
+    # setter
+    self.class_eval("def #{name}=(value);
+      if (value != 'x64' && value != 'x86')
+        raise ArgumentError, \"Invalid data for ProcessorArchitecture.\"
+      end
+      @#{name}=value
+    end")
+  end
+end
+
+class AgentTopologyRequestOperatingSystemTelemetry < StrongTypedClass
+
+  strongtyped_accessor :PercentUserTime, Integer
+  strongtyped_accessor :PercentPrivilegedTime, Integer
+  strongtyped_accessor :UsedMemory, Integer
+  strongtyped_accessor :PercentUsedMemory, Integer
+end
+
+class AgentTopologyRequestOperatingSystem < StrongTypedClass
+  strongtyped_accessor :Name, String
+  strongtyped_accessor :Manufacturer, String
+  strongtyped_arch     :ProcessorArchitecture
+  strongtyped_accessor :Version, String
+  strongtyped_accessor :InContainer, String
+  strongtyped_accessor :Telemetry, AgentTopologyRequestOperatingSystemTelemetry
+end
+ 
+class AgentTopologyRequest < StrongTypedClass
+   
+  strongtyped_accessor :FullyQualfiedDomainName, String
+  strongtyped_accessor :EntityTypeId, String
+  strongtyped_accessor :AuthenticationCertificate, String
+  strongtyped_accessor :OperatingSystem, AgentTopologyRequestOperatingSystem    
+end
+
+def get_telemetry_data(os, telemetry, os_info)
+  if !File.exist?(os_info) && !File.readable?(os_info)
+    raise ArgumentError, " Unable to read file #{os_info}; telemetry information will not be sent to server"
+  end
+
+  if File.exist?('/root/.dockerenv') || File.exist?('/root/.dockerinit')
+    os.InContainer = "True"
+  else
+    os.InContainer = "False"
+  end
+
+  if ENV['TEST_WORKSPACE_ID'].nil? && ENV['TEST_SHARED_KEY'].nil? && File.exist?('/etc/opt/microsoft/omsagent/conf/omsadmin.conf')
+    process_stats=`/opt/omi/bin/omicli wql root/scx "SELECT PercentUserTime, PercentPrivilegedTime, UsedMemory, PercentUsedMemory FROM SCX_UnixProcessStatisticalInformation where Name like 'omsagent'" | grep =`
+    process_stats.each_line do |line|
+      telemetry.PercentUserTime = line.sub("PercentUserTime=","").strip.to_i if line =~ /PercentUserTime/
+      telemetry.PercentPrivilegedTime = line.sub("PercentPrivilegedTime=", "").strip.to_i if  line =~ /PercentPrivilegedTime/
+      telemetry.UsedMemory = line.sub("UsedMemory=", "").strip.to_i if line =~ / UsedMemory/
+      telemetry.PercentUsedMemory = line.sub("PercentUsedMemory=", "").strip.to_i if line =~ /PercentUsedMemory/
+    end
+  end
+
+  File.open(os_info).each_line do |line|
+    os.Name = line.sub("OSName=","").strip if line =~ /OSName/
+    os.Manufacturer = line.sub("OSManufacturer=","").strip if line =~ /OSManufacturer/
+    os.ProcessorArchitecture = "x64"
+    os.Version = line.sub("OSVersion=","").strip if line =~ /OSVersion/
+  end
+end
+
+def obj_to_hash(obj)
+  hash = {}
+  obj.instance_variables.each { |var|
+    val = obj.instance_variable_get(var)
+    if val.is_a? AgentTopologyRequestOperatingSystemTelemetry
+      telemetry_hash = {"Telemetry" => "", :attributes! => {"Telemetry" => obj_to_hash(val) }}
```

Comment:
what does :attributes! mean? does it mean all the properties in Telemetry class should be serialized as attribute in xml? if so, I'd prefer to investigate the property decorator for ruby similar to the "Attribute" in C#.